src/NServiceBus.Core/PreparedEndpoint.cs

```diff
@@ -0,0 +1,101 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Features;
+    using Installation;
+    using ObjectBuilder;
+    using Settings;
+    using Transport;
+
+    /// <summary>
+    /// A prepared endpoint using an external container.
+    /// </summary>
+    public class PreparedEndpoint
+    {
+        /// <summary>
+        /// Provides access to the endpoints message session 
+        /// Note: The message session is only valid to use once the endpoint have been started.
```

Comment:
Maybe make a note that the instance returned is a singleton?

another altternative would be to make this 

```
Lazy<IMessageSesssion>
```

then it is clear it is lazy initialized and a singleton. I have no preference I just wanted to mention it because it crossed my mind while talking to andreas.

Reply:
Now that I see it I think Lazy<IMessageSession> is better since calling it a provide might lead the user to believe that this is a scoped session (which we do want to provide at some later stage)

So I'd say we change to:

`public Lazy<IMessageSession>MessageSession{get; private set;}`

then when we move to a scoped session we ass

`public Lazy<MessageSessionProvider> MessageSessionProvider{get; private set;}`

that users would then use to

`IServiceCollection.AddTransient(_=>provider.Value.OpenSession());` //IScopedMessageSession

@SzymonPobiega what do you think?

Reply:
I am not sure I like this big

```
IServiceCollection.AddTransient(_=>provider.Value.OpenSession());
```

if we really want to provide the `OpenSession()` API then it should be called by the users explicitly, not through DI magic. I think the object that we register in the DI should be the *root* object for the NSB instance and that it should be a singleton.

Reply:
Ok, fair enough, do you agree with the `Lazy<IMessageSession>` part?

Reply:
👍 for lazy. I'll try to change it.